source/source_integration_test.go

```diff
@@ -0,0 +1,436 @@
+	updatedTestContact, err := updateTestContact(ctx, hubspotClient, secondTestContactID)
+	is.NoErr(err)
+
+	record, err = readWithRetry(ctx, source)
+	is.NoErr(err)
+
+	is.Equal(record.Operation, sdk.OperationUpdate)
+
+	compareTestContactWithRecord(t, is, updatedTestContact, record)
+	err = hubspotClient.Delete(ctx, testResource, secondTestContactID)
+	is.NoErr(err)
+
+	err = source.Teardown(ctx)
+	is.NoErr(err)
+}
+
+func TestSource_Read_failBackoffRetry(t *testing.T) {
+	is := is.New(t)
+
+	// prepare a config, configure and open a new source
+	config := prepareConfig(t, "")
+
+	source := NewSource()
+
+	ctx := context.Background()
+
+	err := source.Configure(ctx, config)
+	is.NoErr(err)
+
+	err = source.Open(ctx, nil)
+	is.NoErr(err)
+
+	// since we didn't insert anything we expect that the source
+	// will return the [sdk.ErrBackoffRetry] error
+	_, err = source.Read(ctx)
+	is.Equal(err, sdk.ErrBackoffRetry)
+}
+
+func TestSource_Read_failInvalidToken(t *testing.T) {
+	is := is.New(t)
+
+	// prepare a config with invalid access token, configure and open a new source
+	config := prepareConfig(t, "invalid")
+
+	source := NewSource()
+
+	ctx := context.Background()
+
+	err := source.Configure(ctx, config)
+	is.NoErr(err)
+
+	// we expect to get a 401 error because the access token we provided is invalid
+	err = source.Open(ctx, nil)
+	is.True(err != nil)
+}
+
+func prepareConfig(t *testing.T, accessToken string) map[string]string {
+	t.Helper()
+
+	if accessToken == "" {
+		if testAccessToken == "" {
+			t.Skip("HUBSPOT_ACCESS_TOKEN env var must be set")
+		}
+
+		accessToken = testAccessToken
+	}
+
+	return map[string]string{
+		config.KeyAccessToken: accessToken,
+		config.KeyResource:    testResource,
+	}
+}
+
+// createTestContact writes a random contact and returns it.
+func createTestContact(ctx context.Context, hubspotClient *hubspot.Client) (map[string]any, error) {
+	testContact := map[string]any{
+		"properties": map[string]any{
+			"email":     gofakeit.Email(),
+			"firstname": gofakeit.FirstName(),
+			"lastname":  gofakeit.LastName(),
+		},
+	}
+
+	err := hubspotClient.Create(ctx, testResource, testContact)
+	if err != nil {
+		return nil, fmt.Errorf("create contact: %w", err)
+	}
+
+	return testContact, nil
+}
+
+// updateTestContact updates a random contact's email, firstname and lastname and returns it.
+func updateTestContact(
+	ctx context.Context,
+	hubspotClient *hubspot.Client,
+	itemID string,
+) (map[string]any, error) {
+	updatedTestContact := map[string]any{
+		"properties": map[string]any{
+			"email":     gofakeit.Email(),
+			"firstname": gofakeit.FirstName(),
+			"lastname":  gofakeit.LastName(),
+		},
+	}
+
+	err := hubspotClient.Update(ctx, testResource, itemID, updatedTestContact)
+	if err != nil {
+		return nil, fmt.Errorf("update contact: %w", err)
+	}
+
+	return updatedTestContact, nil
+}
+
+// waitTestContacts waits until the expected count of contacts will be presented in HubSpot.
+func waitTestContacts(
```

Comment:
Do we really need this method?
IIUC, do you want to wait for all records to be written in the specified order before calling the method?

Reply:
Yeah, it seems that I forgot to remove it after adding the `readWithRetry` function. Thanks for the catch!